sympy/sets/sets.py

```diff
@@ -1605,6 +1699,28 @@ def as_relational(self, symbol):
 
         return And(A_rel, B_rel)
 
+    def _kind(self):
+        try:
+            def _measurement(x):
+                if x is UniversalSet or x is S.UniversalSet:
+                    return float('inf')
+                try:
+                    return len(x)
+                except AttributeError:
+                    return int(x._measure)
```

Comment:
When does `len(x)` raise `AttributeError`? That should be fixed rather than catching the exception here.

Reply:
`len(Interval(1, 2))` throws `AttributeError`. But `Interval(1, 2)._measure` works

Reply:
Neither of len or _measure should be used here

Reply:
`Complement({1}, {1, 2, 3}).kind` being `SetKind()` will depend upon first set being smaller in size than second set. This logic is useful when `evaluate=False` because it never simply the set in this process

Reply:
The length doesn't really tell us anything about whether the set is empty or not. This should just always return the kind of the first arg.

Yes, this is inconsistent with what EmptySet does and sometimes a Complement represents the EmptySet. Sometimes it's impossible to know if the Complement represents the EmptySet though:
```python
In [1]: s = FiniteSet(x) - FiniteSet(y)

In [2]: s
Out[2]: {x} \ {y}

In [3]: s.subs({x: 1, y: 1})
Out[3]: ∅

In [4]: s.subs({x: 1, y: 2})
Out[4]: {1}
```
Ideally EmptySet would have a particular kind for its elements but the problem is there's no way for us to choose what that kind should be so it has to be just `SetKind()`. In the case of `Complement` we at least know what the element kind would be if the `Complement` wasn't empty.

Reply:
```python
def _kind(self):
        if type(self.args[0]).__name__ != "ImageSet":
            return self.args[0].kind
        else:
            return SetKind(self.args[0].lamda.expr.kind)
```
This one ok?

Reply:
Types should not be checked by comparing strings.

Anyway this should always just be `return self.args[0].kind`. If that doesn't work for `ImageSet` then it suggests that the kind of `ImageSet` is wrong.

Isn't what you've shown just the way that the kind of `ImageSet` is computed anyway?

Reply:
```python
>>> f = ImageSet(Lambda(n, n ** 2), Interval(1, 4))
>>> type(f).__name__
'ImageSet'
>>> f = ImageSet(Lambda(n, n ** 2), Interval(1, 4)) - FiniteSet(3)
>>> type(f).__name__
'Complement'
```

Reply:
```python
>>> f = ImageSet(Lambda(n, n ** 2), Interval(1, 4))
>>> type(f).__name__
'ImageSet'
>>> f = ImageSet(Lambda(n, n ** 2), Interval(1, 4)) - FiniteSet(3)
>>> type(f).__name__
'Complement'
>>> f.args
(ImageSet(Lambda(n, n**2), Interval(1, 4)), {3})
```
So `return self.args[0].kind` will give error. only `return SetKind(self.args[0].lamda.expr.kind` can give result. I

Reply:
```python
    def _kind(self):
        from .fancysets import ImageSet
        if not isinstance(self.args[0], ImageSet):
            return self.args[0].kind
        else:
            return SetKind(self.args[0].lamda.expr.kind)
```
Will this work

Reply:
> So `return self.args[0].kind` will give error.

I don't see any error:
```python
In [13]: self = ImageSet(Lambda(n, n ** 2), Interval(1, 4)) - FiniteSet(3)

In [14]: self
Out[14]: 
⎧ 2 │           ⎫      
⎨n  │ n ∊ [1, 4]⎬ \ {3}
⎩   │           ⎭      

In [15]: self.args[0].kind
Out[15]: SetKind(NumberKind)
```